File: packages/sdk/src/webhooks.ts
import type {
  CreateWebhookEndpointInput,
  GameId,
  JunjoEvent,
  JunjoEventType,
  Page,
  UpdateWebhookEndpointInput,
  WebhookEndpoint,
  WebhookEndpointFormat,
  WebhookEndpointId,
  WebhookEndpointWithSecret,
  WebhookSignatureHeaders,
} from "@junjo.io/shared";
import { JunjoError } from "./errors.js";
import { UNKNOWN_EVENT_TYPE, type WireJunjoEvent, deserializeEvent } from "./events.js";
import type { HttpClient } from "./http.js";
import { paginate } from "./pagination.js";
import { parseWireDate } from "./wire.js";

/**
 * Signature scheme version prefixed onto every signature. Must stay in
 * sync with the signing layout in the server's `webhookWorker.ts`;
 * bumping one without the other breaks every receiver. Web Crypto is
 * used here (rather than `node:crypto`) so the SDK stays portable
 * across Node 20+ (the supported engines range) and modern browsers
 * without pulling in `@types/node`.
 */
export const WEBHOOK_SIGNATURE_SCHEME = "v1";
/** Default clock-skew tolerance for webhook verification (5 minutes). */
export const WEBHOOK_DEFAULT_TOLERANCE_MS = 5 * 60_000;

const SIGNATURE_HEADER = "x-junjo-signature";
const TIMESTAMP_HEADER = "x-junjo-timestamp";
const EVENT_ID_HEADER = "x-junjo-event-id";
const DELIVERY_ID_HEADER = "x-junjo-delivery-id";

/** Options for webhook verification. */
export interface VerifyOptions {
  /**
   * Maximum allowed clock skew, in milliseconds. Set higher only if
   * your receiver and Junjo's clock are known to drift.
   */
  tolerance?: number;
  /** Override the wall clock for tests. */
  now?: () => Date;
}

/**
 * Any headers object verification accepts: the typed signature headers
 * or a framework-style record (string, string[], or missing values).
 * Header names are matched case-insensitively.
 */
export type WebhookHeaders =
  | WebhookSignatureHeaders
  | Record<string, string | string[] | undefined>;

/** Structural request shape the middleware needs; matches Express. */
export interface ExpressLikeRequest {
  headers: Record<string, string | string[] | undefined>;
  body: unknown;
  rawBody?: Uint8Array | string;
}

/** Structural response shape the middleware needs; matches Express. */
export interface ExpressLikeResponse {
  status(code: number): ExpressLikeResponse;
  send(body?: unknown): void;
  sendStatus(code: number): void;
}

/** The middleware function shape returned by `WebhooksApi.middleware`. */
export type ExpressLikeMiddleware = (
  req: ExpressLikeRequest,
  res: ExpressLikeResponse,
  next: (err?: unknown) => void,
) => Promise<void>;

function pickHeader(headers: WebhookHeaders, name: string): string | undefined {
  const lower = name.toLowerCase();
  for (const [key, value] of Object.entries(headers)) {
    if (key.toLowerCase() !== lower) continue;
    if (typeof value === "string") return value;
    if (Array.isArray(value)) return value[0];
  }
  return undefined;
}

function bodyToString(body: string | Uint8Array): string {
  if (typeof body === "string") return body;
  return new TextDecoder("utf-8").decode(body);
}

function bytesToHex(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length; i++) {
    out += (bytes[i] ?? 0).toString(16).padStart(2, "0");
  }
  return out;
}

async function hmacSha256Hex(secret: string, message: string): Promise<string> {
  const encoder = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const sig = await crypto.subtle.sign("HMAC", key, encoder.encode(message));
  return bytesToHex(new Uint8Array(sig));
}

/**
 * HMAC-SHA256 of `<timestamp>.<body>`, hex, prefixed with the scheme
 * version. Mirrors the server's `signWebhookBody`.
 */
export async function signWebhookBody(
  secret: string,
  body: string,
  timestamp: string,
): Promise<string> {
  const sig = await hmacSha256Hex(secret, `${timestamp}.${body}`);
  return `${WEBHOOK_SIGNATURE_SCHEME}=${sig}`;
}

// Compare two strings of equal length without short-circuiting on the
// first byte that differs. Web Crypto has no public timingSafeEqual.
function constantTimeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

/**
 * Verifies a webhook delivery's signature and timestamp against the
 * endpoint secret and returns the deserialized event. Throws
 * `JunjoError` with a `webhook_*` code on any verification failure.
 * Pass the RAW request body exactly as received; re-serialized JSON
 * will not match the signature.
 */
export async function verifyWebhook(
  rawBody: string | Uint8Array,
  headers: WebhookHeaders,
  secret: string,
  opts: VerifyOptions = {},
): Promise<JunjoEvent> {
  const payload = await verifyWebhookBody(rawBody, headers, secret, opts);
  return deserializeEvent(payload as WireJunjoEvent);
}

/**
 * The signature, timestamp, and JSON checks shared by every verifier,
 * stopping short of event deserialization. Returns the parsed body
 * verbatim; the `unknown` return is deliberate, nothing about the
 * payload shape has been validated beyond it being the JSON the server
 * signed.
 */
async function verifyWebhookBody(
  rawBody: string | Uint8Array,
  headers: WebhookHeaders,
  secret: string,
  opts: VerifyOptions,
): Promise<unknown> {
  const tolerance = opts.tolerance ?? WEBHOOK_DEFAULT_TOLERANCE_MS;
  const now = opts.now ?? (() => new Date());

  const signatureHeader = pickHeader(headers, SIGNATURE_HEADER);
  if (!signatureHeader) {
    throw new JunjoError(`missing ${SIGNATURE_HEADER} header`, "webhook_signature_missing", 400);
  }

  const timestampHeader = pickHeader(headers, TIMESTAMP_HEADER);
  if (!timestampHeader) {
    throw new JunjoError(`missing ${TIMESTAMP_HEADER} header`, "webhook_timestamp_missing", 400);
  }

  // The MAC is checked before the timestamp is even parsed: the
  // signature covers `<timestamp>.<body>` as raw strings, so this needs
  // no parsing, and it means unauthenticated senders get the same
  // `webhook_invalid_signature` for every probe instead of an oracle on
  // the receiver's clock and tolerance settings.
  const body = bodyToString(rawBody);
  const expected = await signWebhookBody(secret, body, timestampHeader);
  if (!constantTimeEqual(signatureHeader, expected)) {
    throw new JunjoError("webhook signature does not match", "webhook_invalid_signature", 400);
  }

  // Date.parse is lenient (it accepts more than ISO 8601), but the
  // timestamp is already authenticated at this point; the server only
  // ever signs ISO strings, so leniency here cannot loosen verification.
  const timestampMs = Date.parse(timestampHeader);
  if (Number.isNaN(timestampMs)) {
    throw new JunjoError(
      `${TIMESTAMP_HEADER} is not a valid timestamp`,
      "webhook_timestamp_invalid",
      400,
    );
  }

  const skew = Math.abs(now().getTime() - timestampMs);
  if (skew > tolerance) {
    throw new JunjoError(
      `signature timestamp is outside the ${tolerance}ms tolerance window`,
      "webhook_timestamp_out_of_tolerance",
      400,
    );
  }

  try {
    return JSON.parse(body);
  } catch {
    throw new JunjoError("webhook body is not valid JSON", "webhook_invalid_body", 400);
  }
}

/** Result of {@link verifyWebhookWithMeta}. */
export interface VerifiedWebhook {
  event: JunjoEvent;
  /**
   * From x-junjo-event-id: stable per event, shared by every delivery
   * (and retry) of that event. The right key for dedupe.
   */
  eventId: string | undefined;
  /** From x-junjo-delivery-id: unique per delivery attempt. */
  deliveryId: string | undefined;
}

/**
 * Result of {@link verifyWebhookWithMeta} under `onUnknownType: "raw"`
 * when the delivery's event type is not in this SDK version's union.
 * The signature and timestamp checks passed; only deserialization was
 * skipped. `event: null` is the discriminant against
 * {@link VerifiedWebhook}.
 */
export interface UnknownVerifiedWebhook {
  event: null;
  /** The wire event type string, when the payload carries one. */
  eventType: string | undefined;
  /** The parsed body verbatim, shape unvalidated. */
  payload: unknown;
  eventId: string | undefined;
  deliveryId: string | undefined;
}

/** Options for {@link verifyWebhookWithMeta}. */
export interface VerifyWithMetaOptions extends VerifyOptions {
  /**
   * What to do when a delivery verifies but carries an event type this
   * SDK version does not know (a newer server). `"throw"` (the default)
   * throws `JunjoError` with code `unknown_event_type`. `"raw"` returns
   * an {@link UnknownVerifiedWebhook} instead, so receivers on match-all
   * endpoints keep acknowledging deliveries across server upgrades.
   */
  onUnknownType?: "throw" | "raw";
}

/**
 * Same verification as `verifyWebhook`, but also surfaces the delivery
 * identity headers so receivers can dedupe retries without re-reading
 * the headers object themselves. Both ids are undefined only when an
 * intermediary stripped the header; the signature check is unchanged.
 */
export async function verifyWebhookWithMeta(
  rawBody: string | Uint8Array,
  headers: WebhookHeaders,
  secret: string,
  opts?: VerifyOptions & { onUnknownType?: "throw" },
): Promise<VerifiedWebhook>;
export async function verifyWebhookWithMeta(
  rawBody: string | Uint8Array,
  headers: WebhookHeaders,
  secret: string,
  opts: VerifyWithMetaOptions,
): Promise<VerifiedWebhook | UnknownVerifiedWebhook>;
export async function verifyWebhookWithMeta(
  rawBody: string | Uint8Array,
  headers: WebhookHeaders,
  secret: string,
  opts: VerifyWithMetaOptions = {},
): Promise<VerifiedWebhook | UnknownVerifiedWebhook> {
  const payload = await verifyWebhookBody(rawBody, headers, secret, opts);
  const eventId = pickHeader(headers, EVENT_ID_HEADER);
  const deliveryId = pickHeader(headers, DELIVERY_ID_HEADER);
  try {
    return { event: deserializeEvent(payload as WireJunjoEvent), eventId, deliveryId };
  } catch (err) {
    if (
      opts.onUnknownType === "raw" &&
      err instanceof JunjoError &&
      err.code === UNKNOWN_EVENT_TYPE
    ) {
      const type = (payload as { type?: unknown } | null)?.type;
      return {
        event: null,
        eventType: typeof type === "string" ? type : undefined,
        payload,
        eventId,
        deliveryId,
      };
    }
    throw err;
  }
}

function readMiddlewareBody(req: ExpressLikeRequest): string | Uint8Array | null {
  if (req.rawBody !== undefined) return req.rawBody;
  if (typeof req.body === "string") return req.body;
  if (req.body instanceof Uint8Array) return req.body;
  return null;
}

interface WireWebhookEndpoint {
  id: string;
  gameId: string;
  url: string;
  events: string[];
  format: string;
  createdAt: string;
  disabledAt: string | null;
}

interface WireWebhookEndpointWithSecret extends WireWebhookEndpoint {
  secret: string;
}

function deserializeEndpoint(w: WireWebhookEndpoint): WebhookEndpoint {
  return {
    id: w.id as WebhookEndpointId,
    gameId: w.gameId as GameId,
    url: w.url,
    events: w.events as JunjoEventType[],
    format: w.format as WebhookEndpointFormat,
    createdAt: parseWireDate(w.createdAt, "createdAt"),
    disabledAt: w.disabledAt === null ? null : parseWireDate(w.disabledAt, "disabledAt"),
  };
}

function deserializeEndpointWithSecret(
  w: WireWebhookEndpointWithSecret,
): WebhookEndpointWithSecret {
  return { ...deserializeEndpoint(w), secret: w.secret };
}

/** Webhook endpoint management: create, list, update, delete. */
export class WebhookEndpointsApi {
  constructor(private readonly http: HttpClient) {}

  /**
   * The signing secret is returned exactly once on create; persist it
   * immediately. `list` and `update` never surface it again.
   */
  async create(
    input: CreateWebhookEndpointInput,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<WebhookEndpointWithSecret> {
    const body: Record<string, unknown> = { url: input.url };
    if (input.events !== undefined) body.events = input.events;
    if (input.secret !== undefined) body.secret = input.secret;
    if (input.format !== undefined) body.format = input.format;
    const wire = await this.http.post<WireWebhookEndpointWithSecret>("/v1/webhooks", body, {
      signal: opts?.signal,
      timeoutMs: opts?.timeoutMs,
    });
    return deserializeEndpointWithSecret(wire);
  }

  /**
   * Cursor-paginated (server default limit 50); `nextCursor` is the id
   * of the last item, fed back in as `cursor` for the next page.
   */
  async list(opts?: {
    limit?: number;
    cursor?: string;
    signal?: AbortSignal;
    timeoutMs?: number;
  }): Promise<Page<WebhookEndpoint>> {
    const params = new URLSearchParams();
    if (opts?.limit !== undefined) params.set("limit", String(opts.limit));
    if (opts?.cursor !== undefined) params.set("cursor", opts.cursor);
    const qs = params.toString();
    const path = qs ? `/v1/webhooks?${qs}` : "/v1/webhooks";
    const wire = await this.http.get<{
      items: WireWebhookEndpoint[];
      nextCursor: string | null;
    }>(path, { signal: opts?.signal, timeoutMs: opts?.timeoutMs });
    return {
      items: wire.items.map(deserializeEndpoint),
      nextCursor: wire.nextCursor,
    };
  }

  /**
   * Async-iterator wrapper over `list(...)` that walks every page until
   * `nextCursor` is null. Endpoint counts are small in practice, but
   * this keeps the surface symmetric with the other paginated lists.
   */
  listAll(opts?: {
    limit?: number;
    signal?: AbortSignal;
    timeoutMs?: number;
  }): AsyncGenerator<WebhookEndpoint> {
    return paginate((cursor) => this.list({ ...opts, cursor }));
  }

  /**
   * The secret is never surfaced by the response; it is only ever
   * returned by `create`.
   */
  async update(
    id: WebhookEndpointId,
    input: UpdateWebhookEndpointInput,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<WebhookEndpoint> {
    const body: Record<string, unknown> = {};
    if (input.url !== undefined) body.url = input.url;
    if (input.events !== undefined) body.events = input.events;
    if (input.disabled !== undefined) body.disabled = input.disabled;
    if (input.format !== undefined) body.format = input.format;
    const wire = await this.http.patch<WireWebhookEndpoint>(
      `/v1/webhooks/${encodeURIComponent(id)}`,
      body,
      { signal: opts?.signal, timeoutMs: opts?.timeoutMs },
    );
    return deserializeEndpoint(wire);
  }

  /**
   * Hard delete; pending deliveries are cascaded by the database.
   * Calling on a missing id throws `JunjoError` with `code: "not_found"`.
   */
  async delete(
    id: WebhookEndpointId,
    opts?: { signal?: AbortSignal; timeoutMs?: number },
  ): Promise<void> {
    await this.http.delete<void>(`/v1/webhooks/${encodeURIComponent(id)}`, undefined, {
      signal: opts?.signal,
      timeoutMs: opts?.timeoutMs,
    });
  }
}

/**
 * Webhooks: endpoint management under `endpoints`, plus signature
 * verification helpers bound to the client for convenience.
 */
export class WebhooksApi {
  readonly endpoints: WebhookEndpointsApi;

  constructor(http: HttpClient) {
    this.endpoints = new WebhookEndpointsApi(http);
  }

  /** Verifies a delivery and returns the event. See `verifyWebhook`. */
  verify(
    rawBody: string | Uint8Array,
    headers: WebhookHeaders,
    secret: string,
    opts?: VerifyOptions,
  ): Promise<JunjoEvent> {
    return verifyWebhook(rawBody, headers, secret, opts);
  }

  /**
   * `verify` plus the event/delivery id headers, for dedupe. See
   * `verifyWebhookWithMeta`.
   */
  verifyWithMeta(
    rawBody: string | Uint8Array,
    headers: WebhookHeaders,
    secret: string,
    opts?: VerifyOptions & { onUnknownType?: "throw" },
  ): Promise<VerifiedWebhook>;
  verifyWithMeta(
    rawBody: string | Uint8Array,
    headers: WebhookHeaders,
    secret: string,
    opts: VerifyWithMetaOptions,
  ): Promise<VerifiedWebhook | UnknownVerifiedWebhook>;
  verifyWithMeta(
    rawBody: string | Uint8Array,
    headers: WebhookHeaders,
    secret: string,
    opts: VerifyWithMetaOptions = {},
  ): Promise<VerifiedWebhook | UnknownVerifiedWebhook> {
    return verifyWebhookWithMeta(rawBody, headers, secret, opts);
  }

  /**
   * Express-compatible middleware. Mount AFTER `express.raw({ type: "application/json" })`
   * so `req.body` is a Buffer; the verified `JunjoEvent` replaces `req.body`
   * before `next()` runs. On verification failure responds 400 with a
   * generic message plus the stable error code and does not call `next()`.
   */
  middleware(secret: string, opts?: VerifyOptions): ExpressLikeMiddleware {
    return async (req, res, next) => {
      const body = readMiddlewareBody(req);
      if (body === null) {
        res
          .status(400)
          .send("webhook middleware requires a raw body (use express.raw before this handler)");
        return;
      }
      let event: JunjoEvent;
      try {
        event = await verifyWebhook(body, req.headers, secret, opts);
      } catch (err) {
        // Error messages can carry receiver internals (e.g. the tolerance
        // window), so only the stable code is reflected to the sender.
        // Full detail stays on the thrown JunjoError for direct verify()
        // callers, who own both sides of the exchange.
        const code = err instanceof JunjoError ? err.code : "webhook_verification_failed";
        res.status(400).send(`webhook verification failed (${code})`);
        return;
      }
      req.body = event;
      next();
    };
  }
}

File: packages/sdk-cpp/src/webhooks.cpp
// Junjo.io SDK for C++

#include "junjo/webhooks.hpp"

#include <cctype>
#include <cstdint>
#include <utility>

#include "junjo/error.hpp"

#include "hmac_sha256.hpp"
#include "request_executor.hpp"
#include "url.hpp"
#include "wire.hpp"

namespace junjo {

namespace {

using detail::Json;

[[nodiscard]] Error invalid_config(std::string message) {
  return Error{.code = ErrorCode::InvalidConfig, .message = std::move(message)};
}

// Serializes a format input, rejecting Unknown (which exists only so
// newer server VALUES deserialize; it has no wire spelling).
[[nodiscard]] Result<std::string> format_to_wire(WebhookFormat format) {
  if (format == WebhookFormat::Unknown) {
    return invalid_config(
        "WebhookFormat::Unknown cannot be sent; pick Junjo, Discord, or Slack");
  }
  return std::string(to_string(format));
}

}  // namespace

// ---------------------------------------------------------------------
// WebhookEndpointsApi
// ---------------------------------------------------------------------

WebhookEndpointsApi::WebhookEndpointsApi(
    std::shared_ptr<const detail::RequestExecutor> executor) noexcept
    : executor_(std::move(executor)) {}

Result<WebhookEndpointWithSecret> WebhookEndpointsApi::create(
    const CreateWebhookEndpointInput& input, const RequestOptions& options,
    const CancellationToken& token) const {
  Json body = Json::object();
  body["url"] = input.url;
  // An empty vector is omitted rather than sent: the server stores an
  // omitted list and an explicit [] identically (match-all), so
  // nothing is lost and the body stays minimal like the TS SDK's.
  if (!input.events.empty()) body["events"] = input.events;
  if (input.secret.has_value()) body["secret"] = *input.secret;
  if (input.format.has_value()) {
    Result<std::string> format = format_to_wire(*input.format);
    if (!format.has_value()) return std::move(format).error();
    body["format"] = std::move(format).value();
  }
  return detail::to_value<WebhookEndpointWithSecret>(
      executor_->execute_json("POST", "/v1/webhooks", body, token, options.timeout),
      detail::deserialize_webhook_endpoint_with_secret);
}

Result<Page<WebhookEndpoint>> WebhookEndpointsApi::list(
    const ListWebhookEndpointsOptions& options, const CancellationToken& token) const {
  std::string limit;
  std::vector<std::pair<std::string_view, std::string_view>> params;
  if (options.limit.has_value()) {
    limit = std::to_string(*options.limit);
    params.emplace_back("limit", limit);
  }
  if (options.cursor.has_value()) params.emplace_back("cursor", *options.cursor);

  const std::string path = "/v1/webhooks" + detail::build_query(params);
  return detail::to_page<WebhookEndpoint>(
      executor_->execute_json("GET", path, std::nullopt, token, options.timeout),
      detail::deserialize_webhook_endpoint);
}

Result<WebhookEndpoint> WebhookEndpointsApi::update(std::string_view id,
                                                    const UpdateWebhookEndpointInput& input,
                                                    const RequestOptions& options,
                                                    const CancellationToken& token) const {
  Json body = Json::object();
  if (input.url.has_value()) body["url"] = *input.url;
  if (input.events.has_value()) body["events"] = *input.events;
  if (input.disabled.has_value()) body["disabled"] = *input.disabled;
  if (input.format.has_value()) {
    Result<std::string> format = format_to_wire(*input.format);
    if (!format.has_value()) return std::move(format).error();
    body["format"] = std::move(format).value();
  }
  const std::string path = "/v1/webhooks/" + detail::percent_encode(id);
  return detail::to_value<WebhookEndpoint>(
      executor_->execute_json("PATCH", path, body, token, options.timeout),
      detail::deserialize_webhook_endpoint);
}

Result<void> WebhookEndpointsApi::remove(std::string_view id, const RequestOptions& options,
                                         const CancellationToken& token) const {
  const std::string path = "/v1/webhooks/" + detail::percent_encode(id);
  return detail::to_void(
      executor_->execute_json("DELETE", path, std::nullopt, token, options.timeout));
}

// ---------------------------------------------------------------------
// WebhooksApi
// ---------------------------------------------------------------------

WebhooksApi::WebhooksApi(std::shared_ptr<const detail::RequestExecutor> executor) noexcept
    : executor_(std::move(executor)) {}

WebhookEndpointsApi WebhooksApi::endpoints() const noexcept {
  return WebhookEndpointsApi(executor_);
}

// ---------------------------------------------------------------------
// Delivery verification
// ---------------------------------------------------------------------

namespace {

[[nodiscard]] Error webhook_error(ErrorCode code, std::string message) {
  return Error{.code = code, .message = std::move(message)};
}

// First header whose name matches ASCII-case-insensitively.
[[nodiscard]] const std::string* pick_header(const WebhookHeaders& headers,
                                             std::string_view name) {
  for (const auto& [key, value] : headers) {
    if (key.size() != name.size()) continue;
    bool matches = true;
    for (std::size_t i = 0; i < name.size(); ++i) {
      const auto a = static_cast<unsigned char>(key[i]);
      const auto b = static_cast<unsigned char>(name[i]);
      if (std::tolower(a) != std::tolower(b)) {
        matches = false;
        break;
      }
    }
    if (matches) return &value;
  }
  return nullptr;
}

// Reads exactly `count` ASCII digits from `text` at `pos` into `out`;
// false when anything else (including end of input) is found.
[[nodiscard]] bool read_digits(std::string_view text, std::size_t& pos, int count,
                               std::int64_t& out) {
  std::int64_t value = 0;
  for (int i = 0; i < count; ++i) {
    if (pos >= text.size() || text[pos] < '0' || text[pos] > '9') return false;
    value = value * 10 + (text[pos] - '0');
    ++pos;
  }
  out = value;
  return true;
}

// Days from 1970-01-01 for a civil date (Howard Hinnant's
// days_from_civil algorithm; public-domain arithmetic).
[[nodiscard]] std::int64_t days_from_civil(std::int64_t year, std::int64_t month,
                                           std::int64_t day) {
  year -= month <= 2 ? 1 : 0;
  const std::int64_t era = (year >= 0 ? year : year - 399) / 400;
  const std::int64_t yoe = year - era * 400;
  const std::int64_t doy = (153 * (month + (month > 2 ? -3 : 9)) + 2) / 5 + day - 1;
  const std::int64_t doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
  return era * 146097 + doe - 719468;
}

[[nodiscard]] bool is_leap_year(std::int64_t year) {
  return (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
}

[[nodiscard]] std::int64_t days_in_month(std::int64_t year, std::int64_t month) {
  constexpr std::int64_t kDays[12] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
  if (month == 2 && is_leap_year(year)) return 29;
  return kDays[month - 1];
}

// Parses an ISO 8601 date-time with an explicit UTC designator or
// numeric offset ("2026-04-28T05:00:30.000Z", offsets as "+hh:mm" /
// "-hh:mm") into milliseconds since the Unix epoch. Fractional seconds
// beyond milliseconds are truncated. Anything else -> nullopt. The
// strictness is safe: this only ever sees an ALREADY AUTHENTICATED
// timestamp, and the server exclusively signs toISOString output, so
// no legitimate delivery is rejected; forms without an explicit offset
// are refused rather than guessed at.
[[nodiscard]] std::optional<std::int64_t> parse_iso8601_ms(std::string_view text) {
  std::size_t pos = 0;
  std::int64_t year = 0;
  std::int64_t month = 0;
  std::int64_t day = 0;
  std::int64_t hour = 0;
  std::int64_t minute = 0;
  std::int64_t second = 0;
  if (!read_digits(text, pos, 4, year)) return std::nullopt;
  if (pos >= text.size() || text[pos] != '-') return std::nullopt;
  ++pos;
  if (!read_digits(text, pos, 2, month)) return std::nullopt;
  if (pos >= text.size() || text[pos] != '-') return std::nullopt;
  ++pos;
  if (!read_digits(text, pos, 2, day)) return std::nullopt;
  if (pos >= text.size() || (text[pos] != 'T' && text[pos] != 't')) return std::nullopt;
  ++pos;
  if (!read_digits(text, pos, 2, hour)) return std::nullopt;
  if (pos >= text.size() || text[pos] != ':') return std::nullopt;
  ++pos;
  if (!read_digits(text, pos, 2, minute)) return std::nullopt;
  if (pos >= text.size() || text[pos] != ':') return std::nullopt;
  ++pos;
  if (!read_digits(text, pos, 2, second)) return std::nullopt;

  std::int64_t milliseconds = 0;
  if (pos < text.size() && text[pos] == '.') {
    ++pos;
    int digits = 0;
    while (pos < text.size() && text[pos] >= '0' && text[pos] <= '9') {
      if (digits < 3) {
        milliseconds = milliseconds * 10 + (text[pos] - '0');
      }
      ++digits;
      ++pos;
    }
    if (digits == 0 || digits > 9) return std::nullopt;
    // Fewer than three digits scale up ("*.5" is 500 ms).
    for (int i = digits; i < 3; ++i) milliseconds *= 10;
  }

  std::int64_t offset_minutes = 0;
  if (pos < text.size() && (text[pos] == 'Z' || text[pos] == 'z')) {
    ++pos;
  } else if (pos < text.size() && (text[pos] == '+' || text[pos] == '-')) {
    const bool negative = text[pos] == '-';
    ++pos;
    std::int64_t offset_hours = 0;
    std::int64_t offset_mins = 0;
    if (!read_digits(text, pos, 2, offset_hours)) return std::nullopt;
    if (pos >= text.size() || text[pos] != ':') return std::nullopt;
    ++pos;
    if (!read_digits(text, pos, 2, offset_mins)) return std::nullopt;
    if (offset_hours > 23 || offset_mins > 59) return std::nullopt;
    offset_minutes = offset_hours * 60 + offset_mins;
    if (negative) offset_minutes = -offset_minutes;
  } else {
    // No offset: refuse rather than guess a zone.
    return std::nullopt;
  }
  if (pos != text.size()) return std::nullopt;

  if (month < 1 || month > 12) return std::nullopt;
  if (day < 1 || day > days_in_month(year, month)) return std::nullopt;
  if (hour > 23 || minute > 59 || second > 59) return std::nullopt;

  const std::int64_t days = days_from_civil(year, month, day);
  const std::int64_t utc_seconds =
      days * 86400 + hour * 3600 + minute * 60 + second - offset_minutes * 60;
  return utc_seconds * 1000 + milliseconds;
}

}  // namespace

std::string sign_webhook_body(std::string_view secret, std::string_view body,
                              std::string_view timestamp) {
  std::string message;
  message.reserve(timestamp.size() + 1 + body.size());
  message.append(timestamp);
  message.push_back('.');
  message.append(body);
  const std::array<std::uint8_t, 32> mac = detail::hmac_sha256(secret, message);
  std::string signature(kWebhookSignatureScheme);
  signature.push_back('=');
  signature += detail::to_hex(mac.data(), mac.size());
  return signature;
}

Result<VerifiedWebhook> verify_webhook(std::string_view raw_body, const WebhookHeaders& headers,
                                       std::string_view secret,
                                       const VerifyWebhookOptions& options) {
  const std::string* signature = pick_header(headers, kWebhookSignatureHeader);
  if (signature == nullptr) {
    return webhook_error(ErrorCode::WebhookSignatureMissing,
                         "missing x-junjo-signature header");
  }
  const std::string* timestamp = pick_header(headers, kWebhookTimestampHeader);
  if (timestamp == nullptr) {
    return webhook_error(ErrorCode::WebhookTimestampMissing,
                         "missing x-junjo-timestamp header");
  }

  // The MAC is checked before the timestamp is even parsed: the
  // signature covers "<timestamp>.<body>" as raw strings, so this
  // needs no parsing, and it means unauthenticated senders get the
  // same WebhookInvalidSignature for every probe instead of an oracle
  // on the receiver's clock and tolerance settings (parity with the TS
  // SDK).
  const std::string expected = sign_webhook_body(secret, raw_body, *timestamp);
  if (!detail::constant_time_equal(*signature, expected)) {
    return webhook_error(ErrorCode::WebhookInvalidSignature,
                         "webhook signature does not match");
  }

  const std::optional<std::int64_t> timestamp_ms = parse_iso8601_ms(*timestamp);
  if (!timestamp_ms.has_value()) {
    return webhook_error(ErrorCode::WebhookTimestampInvalid,
                         "x-junjo-timestamp is not a valid timestamp");
  }

  const std::chrono::system_clock::time_point now =
      options.now ? options.now() : std::chrono::system_clock::now();
  const std::int64_t now_ms =
      std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()).count();
  const std::int64_t skew_ms =
      now_ms >= *timestamp_ms ? now_ms - *timestamp_ms : *timestamp_ms - now_ms;
  if (skew_ms > options.tolerance.count()) {
    return webhook_error(ErrorCode::WebhookTimestampOutOfTolerance,
                         "signature timestamp is outside the tolerance window");
  }

  const std::optional<Json> parsed = detail::parse_json(raw_body);
  if (!parsed.has_value() || !parsed->is_object()) {
    return webhook_error(ErrorCode::WebhookInvalidBody, "webhook body is not a JSON object");
  }
  const auto type = parsed->find("type");
  if (type == parsed->end() || !type->is_string()) {
    return webhook_error(ErrorCode::WebhookInvalidBody,
                         "webhook body is missing string field type");
  }

  VerifiedWebhook verified;
  verified.event_type = type->get<std::string>();
  if (const std::string* event_id = pick_header(headers, kWebhookEventIdHeader)) {
    verified.event_id = *event_id;
  }
  if (const std::string* delivery_id = pick_header(headers, kWebhookDeliveryIdHeader)) {
    verified.delivery_id = *delivery_id;
  }
  verified.payload_json = std::string(raw_body);
  return verified;
}

}  // namespace junjo
